File: backend/src/models/direccion.rs
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use chrono::NaiveDateTime;
use std::str::FromStr;

// ==================== ENUMS ====================

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "tipo_direccion", rename_all = "lowercase")]
pub enum TipoDireccion {
    #[sqlx(rename = "envio")]
    Envio,
    #[sqlx(rename = "facturacion")]
    Facturacion,
    #[sqlx(rename = "ambos")]
    Ambos,
}

impl FromStr for TipoDireccion {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "envio" => Ok(TipoDireccion::Envio),
            "facturacion" => Ok(TipoDireccion::Facturacion),
            "ambos" => Ok(TipoDireccion::Ambos),
            _ => Err(format!("Tipo de dirección inválido: {}", s)),
        }
    }
}

// ==================== MODELO PRINCIPAL ====================

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct Direccion {
    pub id_direccion: i32,
    pub id_usuario: i32,
    pub tipo: Option<String>,
    pub nombre_completo: Option<String>,
    pub direccion_linea1: String,
    pub direccion_linea2: Option<String>,
    pub ciudad: String,
    pub departamento: String,
    pub codigo_postal: Option<String>,
    pub pais: Option<String>,
    pub telefono_contacto: Option<String>,
    pub referencia: Option<String>,
    pub es_predeterminada: Option<bool>,
    pub activo: Option<bool>,
    pub fecha_creacion: Option<NaiveDateTime>,
}

// ==================== DTOs DE RESPUESTA ====================

#[derive(Debug, Serialize)]
pub struct DireccionResponse {
    pub id_direccion: i32,
    pub tipo: String,
    pub nombre_completo: Option<String>,
    pub direccion_linea1: String,
    pub direccion_linea2: Option<String>,
    pub ciudad: String,
    pub departamento: String,
    pub codigo_postal: Option<String>,
    pub pais: Option<String>,
    pub telefono_contacto: Option<String>,
    pub referencia: Option<String>,
    pub es_predeterminada: bool,
}

impl From<Direccion> for DireccionResponse {
    fn from(d: Direccion) -> Self {
        DireccionResponse {
            id_direccion: d.id_direccion,
            tipo: d.tipo.unwrap_or_else(|| "envio".to_string()),
            nombre_completo: d.nombre_completo,
            direccion_linea1: d.direccion_linea1,
            direccion_linea2: d.direccion_linea2,
            ciudad: d.ciudad,
            departamento: d.departamento,
            codigo_postal: d.codigo_postal,
            pais: d.pais,
            telefono_contacto: d.telefono_contacto,
            referencia: d.referencia,
            es_predeterminada: d.es_predeterminada.unwrap_or(false),
        }
    }
}

// ==================== DTOs DE REQUEST ====================

#[derive(Debug, Deserialize)]
pub struct CrearDireccionRequest {
    pub tipo: Option<String>, // "envio", "facturacion", "ambos"
    pub nombre_completo: Option<String>,
    pub direccion_linea1: String,
    pub direccion_linea2: Option<String>,
    pub ciudad: String,
    pub departamento: String,
    pub codigo_postal: Option<String>,
    pub pais: Option<String>, // Default: "Perú"
    pub telefono_contacto: Option<String>,
    pub referencia: Option<String>,
    pub es_predeterminada: Option<bool>,
}

#[derive(Debug, Deserialize)]
pub struct ActualizarDireccionRequest {
    pub tipo: Option<String>,
    pub nombre_completo: Option<String>,
    pub direccion_linea1: Option<String>,
    pub direccion_linea2: Option<String>,
    pub ciudad: Option<String>,
    pub departamento: Option<String>,
    pub codigo_postal: Option<String>,
    pub pais: Option<String>,
    pub telefono_contacto: Option<String>,
    pub referencia: Option<String>,
    pub es_predeterminada: Option<bool>,
}

File: frontend/src/lib/services/direccion.ts
import { apiAuth } from './auth';

// ==================== INTERFACES ====================

export interface Direccion {
	id_direccion: number;
	tipo: 'envio' | 'facturacion' | 'ambos';
	nombre_completo?: string;
	direccion_linea1: string;
	direccion_linea2?: string;
	ciudad: string;
	departamento: string;
	codigo_postal?: string;
	pais: string;
	telefono_contacto?: string;
	referencia?: string;
	es_predeterminada: boolean;
	activo: boolean;
	fecha_creacion: string;
}

export interface CrearDireccionRequest {
	tipo: 'envio' | 'facturacion' | 'ambos';
	nombre_completo?: string;
	direccion_linea1: string;
	direccion_linea2?: string;
	ciudad: string;
	departamento: string;
	codigo_postal?: string;
	pais?: string; // Default: "Perú"
	telefono_contacto?: string;
	referencia?: string;
	es_predeterminada?: boolean;
}

export interface ActualizarDireccionRequest {
	tipo?: 'envio' | 'facturacion' | 'ambos';
	nombre_completo?: string;
	direccion_linea1?: string;
	direccion_linea2?: string;
	ciudad?: string;
	departamento?: string;
	codigo_postal?: string;
	pais?: string;
	telefono_contacto?: string;
	referencia?: string;
	es_predeterminada?: boolean;
	activo?: boolean;
}

interface ApiResponse<T> {
	success: boolean;
	data?: T;
	message?: string;
}

// ==================== SERVICIO ====================

class DireccionService {
	/**
	 * Obtener todas las direcciones del usuario
	 */
	async getDirecciones(): Promise<Direccion[]> {
		try {
			const { data } = await apiAuth.get<ApiResponse<Direccion[]>>('/direcciones');
			if (data.success && data.data) {
				return data.data;
			}
			throw new Error(data.message || 'Error al obtener direcciones');
		} catch (error: any) {
			console.error('Error en getDirecciones:', error);
			throw new Error(error.response?.data?.message || 'Error al obtener direcciones');
		}
	}

	/**
	 * Crear una nueva dirección
	 */
	async crearDireccion(request: CrearDireccionRequest): Promise<Direccion> {
		try {
			const { data } = await apiAuth.post<ApiResponse<Direccion>>('/direcciones', request);
			if (data.success && data.data) {
				return data.data;
			}
			throw new Error(data.message || 'Error al crear dirección');
		} catch (error: any) {
			console.error('Error en crearDireccion:', error);
			throw new Error(error.response?.data?.message || 'Error al crear dirección');
		}
	}

	/**
	 * Actualizar una dirección existente
	 */
	async actualizarDireccion(id: number, request: ActualizarDireccionRequest): Promise<Direccion> {
		try {
			const { data } = await apiAuth.put<ApiResponse<Direccion>>(
				`/direcciones/${id}`,
				request
			);
			if (data.success && data.data) {
				return data.data;
			}
			throw new Error(data.message || 'Error al actualizar dirección');
		} catch (error: any) {
			console.error('Error en actualizarDireccion:', error);
			throw new Error(error.response?.data?.message || 'Error al actualizar dirección');
		}
	}

	/**
	 * Eliminar una dirección
	 */
	async eliminarDireccion(id: number): Promise<void> {
		try {
			const { data } = await apiAuth.delete<ApiResponse<null>>(`/direcciones/${id}`);
			if (!data.success) {
				throw new Error(data.message || 'Error al eliminar dirección');
			}
		} catch (error: any) {
			console.error('Error en eliminarDireccion:', error);
			throw new Error(error.response?.data?.message || 'Error al eliminar dirección');
		}
	}
}

export const direccionService = new DireccionService();
